retry/retry.go

```diff
@@ -30,61 +30,124 @@ type OnSuccessFunc func(elapsed time.Duration, attempt uint64, lastErr error)
 
 // Settings aggregates optional settings for WithBackoff.
 type Settings struct {
-	// If >0, Timeout lets WithBackoff stop retrying gracefully once elapsed based on the following criteria:
-	// * If the execution of RetryableFunc has taken longer than Timeout, no further attempts are made.
-	// * If Timeout elapses during the sleep phase between retries, one final retry is attempted.
-	// * RetryableFunc is always granted its full execution time and is not canceled if it exceeds Timeout.
+	// Timeout, if > 0, lets WithBackoff stop retrying gracefully once elapsed based on the following criteria:
+	//
+	// 	* If the execution of RetryableFunc has taken longer than Timeout, no further attempts are made.
+	// 	* If Timeout elapses during the sleep phase between retries, one final retry is attempted.
+	// 	* RetryableFunc is always granted its full execution time and is not canceled if it exceeds Timeout.
+	//
 	// This means that WithBackoff may not stop exactly after Timeout expires,
 	// or may not retry at all if the first execution of RetryableFunc already takes longer than Timeout.
-	Timeout          time.Duration
+	Timeout time.Duration
+
+	// OnRetryableError, if not nil, is called if a retryable error occurred.
 	OnRetryableError OnRetryableErrorFunc
-	OnSuccess        OnSuccessFunc
+
+	// OnSuccess, if not nil, is called after the function succeeded.
+	OnSuccess OnSuccessFunc
+
+	// QuickContextExit, if set, directly aborts if the context is done and does not wait for functions to finish.
+	QuickContextExit bool
+}
+
+// contextBoundFunc wraps an error generating function, but directly exits with an error if the context is done.
+//
+// While the function parameter signature matches RetryableFunc, this is not used here as this function does not only
+// focuses on RetryableFunc, but more generally functions which might get aborted within WithBackoff.
+func contextBoundFunc(f func(context.Context) error) func(context.Context) error {
+	return func(ctx context.Context) error {
+		fResultCh := make(chan error)
+
+		go func() {
+			fResultCh <- f(ctx)
+			close(fResultCh)
+		}()
+
+		select {
+		case err := <-fResultCh:
+			return err
+
+		case <-ctx.Done():
+			go func() {
+				// Drain channel and wait for it to be closed in the background.
+				for range fResultCh {
+				}
+			}()
+
+			return ctx.Err()
+		}
+	}
 }
 
-// WithBackoff retries the passed function if it fails and the error allows it to retry.
-// The specified backoff policy is used to determine how long to sleep between attempts.
+// WithBackoff retries the retryableFunc in case of failure.
+//
+// The passed ctx loosely restricts retries and prohibits retries if the context is done. It is also being passed to the
+// retryableFunc, which should honor the context as well. However, unless Settings.QuickContextExit is set, the
+// RetryableFunc can survive after the context has already finished.
```

Comment:
Doesn't it also survive if `QuickContextExit` is set, just with the difference that it's running in a different goroutine in the background?